pyquil/api/_quantum_computer.py

```diff
@@ -154,40 +131,87 @@ def run(self, executable: Executable,
+        :param meas_qubits: An advanced feature. The groups of measurement qubits. Only these
+            qubits will be symmetrized over, even if the program acts on other qubits.
+        :return: A numpy array of shape (trials, len(ro-register)) that contains 0s and 1s.
         """
-        Run a quil program in such a way that the readout error is made collectively symmetric
+        if type(symm_type) is not int:
+            raise ValueError("Symmetrization options are indicated by an int. See "
+                             "the docstrings for more information.")
 
-        This means the probability of a bitstring ``b`` being mistaken for a bitstring ``c`` is
-        the same as the probability of ``not(b)`` being mistaken for ``not(c)``
+        if meas_qubits is None:
+            meas_qubits = list(program.get_qubits())
 
-        A more general symmetrization would guarantee that the probability of ``b`` being
-        mistaken for ``c`` depends only on which bit of ``c`` are different from ``b``. This
-        would require choosing random subsets of bits to flip.
+        # It is desirable to have hundreds or thousands of trials more than the minimum
+        trials = _check_min_num_trials_for_symmetrized_readout(len(meas_qubits), trials, symm_type)
 
-        In a noisy device, the probability of accurately reading the 0 state might be higher
-        than that of the 1 state. This makes correcting for readout more difficult. This
-        function runs the program normally ``(trials//2)`` times. The other half of the time,
-        it will insert an ``X`` gate prior to any ``MEASURE`` instruction and then flip the
-        measured classical bit back.
+        sym_programs, flip_arrays = _symmetrization(program, meas_qubits, symm_type)
 
-        See :py:func:`run` for this function's parameter descriptions.
-        """
-        flipped_program = _get_flipped_protoquil_program(program)
-        if trials % 2 != 0:
-            raise ValueError("Using symmetrized measurement functionality requires that you "
-                             "take an even number of trials.")
-        half_trials = trials // 2
-        flipped_program = flipped_program.wrap_in_numshots_loop(shots=half_trials)
-        flipped_executable = self.compile(flipped_program)
-
-        executable = self.compile(program.wrap_in_numshots_loop(half_trials))
-        samples = self.run(executable)
-        flipped_samples = self.run(flipped_executable)
-        double_flipped_samples = np.logical_not(flipped_samples).astype(int)
-        results = np.concatenate((samples, double_flipped_samples), axis=0)
-        np.random.shuffle(results)
-        return results
+        # Floor division so e.g. 9 // 8 = 1 and 17 // 8 = 2
+        num_shots_per_prog = trials // len(sym_programs)
+
+        if num_shots_per_prog * len(sym_programs) < trials:
+            print("The number of trials was modified from " + str(trials) + " to " + str(
```

Comment:
prefer f-strings, and log statements or warnings over print statements

Reply:
- [x] here todo
- [x] add f string below circa line 1080 todo